src/icon.rs

```diff
@@ -125,11 +141,61 @@ impl fmt::Debug for Icon {
 impl Icon {
     /// Creates an `Icon` from 32bpp RGBA data.
     ///
-    /// The length of `rgba` must be divisible by 4, and `width * height` must equal
-    /// `rgba.len() / 4`. Otherwise, this will return a `BadIcon` error.
-    pub fn from_rgba(rgba: Vec<u8>, width: u32, height: u32) -> Result<Self, BadIcon> {
+    /// ## Panics
+    /// Panics if the length of `rgba` must be divisible by 4, or if `width * height` doesn't equal
+    /// `rgba.len() / 4`.
+    pub fn from_rgba(rgba: &[u8], size: PhysicalSize<u32>) -> Result<Self, io::Error> {
+        Ok(Icon {
+            inner: PlatformIcon::from_rgba(rgba.into(), size)?,
+        })
+    }
+
+    /// Creates an `Icon` from 32bpp RGBA data, with a defined cursor hot spot. The hot spot is
+    /// the exact pixel in the icon image where the cursor clicking point is, and is ignored when
+    /// the icon is used as a window icon.
+    ///
+    /// ## Panics
+    /// Panics if the length of `rgba` must be divisible by 4, or if `width * height` doesn't equal
+    /// `rgba.len() / 4`.
+    pub fn from_rgba_with_hot_spot(
+        rgba: &[u8],
+        size: PhysicalSize<u32>,
+        hot_spot: PhysicalPosition<u32>,
+    ) -> Result<Self, io::Error> {
+        Ok(Icon {
+            inner: PlatformIcon::from_rgba_with_hot_spot(rgba.into(), size, hot_spot)?,
+        })
+    }
+
+    /// Lazily create an icon from several scaled source images.
+    ///
+    /// `get_icon` will be lazily called for a particular icon size whenever the window manager
+    /// needs an icon of that size. The `PhysicalSize<u32>` parameter specifies the window manager's
+    /// suggested icon size for a particular scale factor, and will always be a square. The `f64`
+    /// parameter specifies the scale factor that the window manager is requesting the icon with.
+    /// `get_icon` will only be called once for any given suggested icon size.
+    ///
+    /// If `get_icon` returns `Err(e)` for a given size, Winit will invoke `warn!` on the returned
+    /// error and will try to retrieve a differently-sized icon from `get_icon`.
+    pub fn from_rgba_fn<F, B>(mut get_icon: F) -> Result<Self, std::io::Error>
```

Comment:
I don't really like that the error is `std::io::Error` here, since there's no io involved, the only thing that could happen is that cursor isn't found? In that case, having a clear type or even using `Option` should be better? The option applies to the closure you're using as well, so if something is not present it'll be `None`, not some random io error.

Reply:
`io::Error` is used as the standard OS error type in Rust, even in cases where the error isn't directly related to IO (see [`io::Error::last_os_error()`](https://doc.rust-lang.org/std/io/struct.Error.html#method.last_os_error)). Still, it's not relevant for this function since `get_icon` is supposed to be called lazily, so I've changed the function to return `Self` instead of a `Result`.